streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -792,54 +792,67 @@ trait ZStream[-R, +E, +A] extends Serializable { self =>
     } yield queue
 
   /**
-   * Applies a transducer to the stream, which converts one or more elements
-   * of type `A` into elements of type `C`.
+   * Applies a transducer to the stream, converting elements of type `A` into elements of type `C`, with a
+   * managed resource of type `D` available.
    */
-  final def transduce[R1 <: R, E1 >: E, A1 >: A, C](sink: ZSink[R1, E1, A1, A1, C]): ZStream[R1, E1, C] =
+  final def transduceManaged[R1 <: R, E1 >: E, A1 >: A, C](
+    managedSink: ZManaged[R1, E1, ZSink[R1, E1, A1, A1, C]]
+  ): ZStream[R1, E1, C] =
     new ZStream[R1, E1, C] {
-      override def fold[R2 <: R1, E2 >: E1, C1 >: C, S2]: Fold[R2, E2, C1, S2] =
-        ZManaged.succeedLazy { (s2, cont, f) =>
-          def feed(s1: sink.State, s2: S2, a: Chunk[A1]): ZIO[R2, E2, (sink.State, S2, Boolean)] =
-            sink.stepChunk(s1, a).flatMap { step =>
-              if (ZSink.Step.cont(step)) {
-                IO.succeed((ZSink.Step.state(step), s2, true))
-              } else {
-                sink.extract(ZSink.Step.state(step)).flatMap { c =>
-                  f(s2, c).flatMap { s2 =>
-                    val remaining = ZSink.Step.leftover(step)
-                    sink.initial.flatMap { initStep =>
-                      if (cont(s2) && !remaining.isEmpty) {
-                        feed(ZSink.Step.state(initStep), s2, remaining)
-                      } else {
-                        IO.succeed((ZSink.Step.state(initStep), s2, false))
+      override def fold[R2 <: R1, E2 >: E1, C1 >: C, S]: Fold[R2, E2, C1, S] =
+        managedSink.either.flatMap {
```

Comment:
For consistency with other combinators, let's compose the `ZManaged` value like so:
```scala
ZManaged.succeedLazy { (f, cont, s) =>
  for {
    sink <- managedSink
    fold <- self.fold[...]
    s <- f0(...)
  } yield s
}
```

This should keep the ZManaged scopes composed correctly, because we want to tie the sink allocation to the execution of the fold, not the execution of the trampoline (which is the outer ZManaged layer in the `Fold` type).

This should also save the need for the `either` on the sink.

Reply:
Ok, that makes total sense, done.